File: src/main/java/net/orandja/vw/CustomRecipe.java
package net.orandja.vw;

import com.google.common.collect.Maps;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.inventory.CraftingInventory;
import net.minecraft.item.ItemStack;
import net.minecraft.recipe.CraftingRecipe;
import net.minecraft.recipe.Ingredient;
import net.minecraft.recipe.ShapedRecipe;
import net.minecraft.recipe.ShapelessRecipe;
import net.minecraft.util.Identifier;
import net.minecraft.util.collection.DefaultedList;

import java.util.Map;

public interface CustomRecipe {

    interface Interceptor {
        ItemStack onTakeItem(CraftingInventory input , PlayerEntity player, int slot, int amount);
    }

    interface ShapedRecipeConstructor {
        ShapedRecipe create(Identifier identifier, String group, int width, int height, DefaultedList<Ingredient> input, ItemStack output);
    }

    interface ShapelessRecipeConstructor {
        ShapelessRecipe create(Identifier identifier, String group, ItemStack output, DefaultedList<Ingredient> input);
    }

    Map<Identifier, ShapedRecipeConstructor> customShapedRecipes = Maps.newHashMap();
    Map<Identifier, ShapelessRecipeConstructor> customShapelessRecipes = Maps.newHashMap();

    default boolean hasCustomRecipe(Identifier identifier) {
        return customShapelessRecipes.containsKey(identifier) || customShapedRecipes.containsKey(identifier);
    }

    default ShapedRecipe createShapedRecipe(Identifier identifier, String group, int width, int height, DefaultedList<Ingredient> input, ItemStack output) {
        if(customShapedRecipes.containsKey(identifier)) {
            return customShapedRecipes.get(identifier).create(identifier, group, width, height, input, output);
        }

        return null;
    }

    default ShapedRecipe createFrom(ShapedRecipe recipe) {
        return createShapedRecipe(recipe.getId(), recipe.getGroup(), recipe.getWidth(), recipe.getHeight(), recipe.getIngredients(), recipe.getOutput());
    }

    default ShapelessRecipe createFrom(ShapelessRecipe recipe) {
        return createShapelessRecipe(recipe.getId(), recipe.getGroup(), recipe.getOutput(), recipe.getIngredients());
    }

    default ShapelessRecipe createShapelessRecipe(Identifier identifier, String group, ItemStack output, DefaultedList<Ingredient> input) {
        if(customShapelessRecipes.containsKey(identifier)) {
            return customShapelessRecipes.get(identifier).create(identifier, group, output, input);
        }

        return null;
    }

    default ItemStack interceptOnTakeItem(CraftingRecipe recipe, CraftingInventory input, PlayerEntity player, int slot, int amount) {
        if(recipe instanceof Interceptor interceptor) {
            return ((Interceptor) recipe).onTakeItem(input, player, slot, amount);
        }

        return input.removeStack(slot, amount);
    }

}

File: src/main/kotlin/net/orandja/vw/logic/CustomRecipe.kt
package net.orandja.vw.logic

import net.minecraft.entity.player.PlayerEntity
import net.minecraft.inventory.CraftingInventory
import net.minecraft.item.ItemStack
import net.minecraft.recipe.*
import net.minecraft.util.Identifier
import net.minecraft.util.collection.DefaultedList

interface CustomRecipeInterceptor {

    fun onTakeItem(input: CraftingInventory, player: PlayerEntity, slot: Int, amount: Int): ItemStack

}

interface CustomRecipe {

    companion object {

        val customShapedRecipes: MutableMap<Identifier, (identifier: Identifier, group: String, width: Int, height: Int, input: DefaultedList<Ingredient>, output: ItemStack) -> ShapedRecipe?> = mutableMapOf()
        val customShapelessRecipes: MutableMap<Identifier, (identifier: Identifier, group: String, output: ItemStack, input: DefaultedList<Ingredient>) -> ShapelessRecipe?> = mutableMapOf()

    }

    fun createShapedRecipe(identifier: Identifier, group: String, width: Int, height: Int, input: DefaultedList<Ingredient>, output: ItemStack): ShapedRecipe {
        return customShapedRecipes[identifier]?.invoke(identifier, group, width, height, input, output) ?: ShapedRecipe(identifier, group, width, height, input, output)
    }

    fun createShapelessRecipe(identifier: Identifier, group: String, output: ItemStack, input: DefaultedList<Ingredient>): ShapelessRecipe {
        return customShapelessRecipes[identifier]?.invoke(identifier, group, output, input) ?: ShapelessRecipe(identifier, group, output, input)
    }

    fun interceptOnTakeItem(recipe: CraftingRecipe?, input: CraftingInventory, player: PlayerEntity, slot: Int, amount: Int): ItemStack {
        if(recipe is CustomRecipeInterceptor) {
            return recipe.onTakeItem(input, player, slot, amount)
        }

        return input.removeStack(slot, amount)
    }

}